File: app/src/main/java/app/wt/noolis/adapter/ListAdapterCategory.java
package app.wt.noolis.adapter;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import app.wt.noolis.R;
import app.wt.noolis.model.Category;
import app.wt.noolis.utils.Tools;

public class ListAdapterCategory extends RecyclerView.Adapter<ListAdapterCategory.ViewHolder> {

    private List<Category> items = new ArrayList<>();

    private Context ctx;
    private OnItemClickListener mOnItemClickListener;

    public interface OnItemClickListener {
        void onItemClick(View view, Category obj, int position);
    }

    public void setOnItemClickListener(final OnItemClickListener mItemClickListener) {
        this.mOnItemClickListener = mItemClickListener;
    }

    // Provide a suitable constructor (depends on the kind of dataset)
    public ListAdapterCategory(Context context, List<Category> items) {
        this.items = items;
        ctx = context;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        // each data item is just a string in this case
        public TextView name;
        public ImageView image;
        public TextView counter;
        public LinearLayout lyt_parent;

        public ViewHolder(View v) {
            super(v);
            name = (TextView) v.findViewById(R.id.name);
            counter = (TextView) v.findViewById(R.id.counter);
            image = (ImageView) v.findViewById(R.id.image);
            lyt_parent = (LinearLayout) v.findViewById(R.id.lyt_parent);
        }
    }


    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        // create a new view
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_category, parent, false);

        // set the view's size, margins, paddings and layout parameters
        ViewHolder vh = new ViewHolder(v);
        return vh;
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(ViewHolder holder, final int position) {
        final Category c = items.get(position);
        holder.name.setText(c.getName());
        holder.counter.setText(c.getNote_count()+"");
        holder.image.setImageResource(Tools.StringToResId(c.getIcon(),ctx));
        ((GradientDrawable) holder.image.getBackground()).setColor(Color.parseColor(c.getColor()));
        holder.lyt_parent.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mOnItemClickListener.onItemClick(view, c, position);
            }
        });
    }

    public Category getItem(int position){
        return items.get(position);
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return items.size();
    }

}

File: app/src/main/java/app/wt/noolis/adapter/ListAdapterCategory.kt
package app.wt.noolis.adapter

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import java.util.ArrayList

import app.wt.noolis.R
import app.wt.noolis.model.Category
import app.wt.noolis.utils.Tools

class ListAdapterCategory// Provide a suitable constructor (depends on the kind of dataset)
(private val ctx: Context, items: List<Category>) : RecyclerView.Adapter<ListAdapterCategory.ViewHolder>() {

    private var items = ArrayList<Category>()
    private var mOnItemClickListener: OnItemClickListener? = null

    interface OnItemClickListener {
        fun onItemClick(view: View, obj: Category, position: Int)
    }

    fun setOnItemClickListener(mItemClickListener: (Any, Category, Any) -> Unit) {
        this.mOnItemClickListener = mItemClickListener as OnItemClickListener
    }

    init {
        this.items = items as ArrayList<Category>
    }

    inner class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        // each data item is just a string in this case
        var name: TextView = v.findViewById<View>(R.id.name) as TextView
        var image: ImageView
        var counter: TextView
        var lyt_parent: LinearLayout

        init {
            counter = v.findViewById<View>(R.id.counter) as TextView
            image = v.findViewById<View>(R.id.image) as ImageView
            lyt_parent = v.findViewById<View>(R.id.lyt_parent) as LinearLayout
        }
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        // create a new view
        val v = LayoutInflater.from(parent.context).inflate(R.layout.row_category, parent, false)

        // set the view's size, margins, paddings and layout parameters
        val vh = ViewHolder(v)
        return vh
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val c = items[position]
        holder.name.text = c.name
        holder.counter.text = c.note_count.toString() + ""
        holder.image.setImageResource(Tools.StringToResId(c.icon!!, ctx))
        (holder.image.background as GradientDrawable).setColor(Color.parseColor(c.color))
        holder.lyt_parent.setOnClickListener { view -> mOnItemClickListener!!.onItemClick(view, c, position) }
    }

    fun getItem(position: Int): Category {
        return items[position]
    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {
        return items.size
    }

}